File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.util.Scanner;

public class Main {

    public static int[] parent;

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int m = sc.nextInt();

        parent = new int[n + 1];

        for (int i = 0; i <= n; i++) {
            parent[i] = i;
        }

        for (int i = 0; i < m; i++) {
            int q = sc.nextInt();
            int a = sc.nextInt();
            int b = sc.nextInt();
            if (q == 0) {
                union(a, b);
            } else {
                if (checkSame(a, b)) {
                    System.out.println("YES");
                } else {
                    System.out.println("NO");
                }
            }
        }

    }
    public static void union(int a, int b) {
        a = find(a);
        b = find(b);
        if (a != b) {
            parent[b] = a;
        }
    }

    public static int find(int a) {
        if (a == parent[a]) return a;
        else return parent[a] = find(parent[a]);
    }

    public static boolean checkSame(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) return true;
        else return false;
    }

}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.kt
import java.util.*
import java.io.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())

    val n = st.nextInt()
    val m = st.nextInt()

    val arr = IntArray(n + 1) { it }

    for (i in 0 until m) {
        st = StringTokenizer(br.readLine())

        val operator = st.nextInt()
        val a = st.nextInt()
        val b = st.nextInt()

        if (operator == 0) union(arr, a, b)
        else println(check(arr, a, b))
    }
}

fun union(arr: IntArray, a: Int, b: Int) {
    val rootA = find(arr, a)
    val rootB = find(arr, b)

    if (rootA != rootB) arr[rootB] = rootA
}

fun check(arr: IntArray, a: Int, b: Int): String {
    val rootA = find(arr, a)
    val rootB = find(arr, b)

    return if (rootA == rootB) "YES" else "NO"
}

fun find(arr: IntArray, num: Int): Int {
    val found = arr[num]
    if (found == num) return found

    return find(arr, found).also { arr[num] = it }
}

private fun StringTokenizer.nextInt(): Int = nextToken().toInt()
